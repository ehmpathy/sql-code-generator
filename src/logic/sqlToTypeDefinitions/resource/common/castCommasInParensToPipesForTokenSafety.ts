export const castCommasInParensToPipesForTokenSafety = ({ sql }: { sql: string }) => {
  // track state of whether we are in parenthesizes or not
  let areInsideParensDepth = 0; // init as 0, since not inside parens
  const areInsideParens = () => areInsideParensDepth > 0; // whenever areInsideParensDepth > 0, we are inside parens

  // init object to build up the "casted" string
  let castedString = '';

  // loop through each char and update state / replace char as needed
  [...sql].forEach((char) => {
    // 1. manage "areInsideParens" state
    if (char === '(') {
      areInsideParensDepth += 1; // one parens deeper!
    }
    if (char === ')') {
      if (!areInsideParens) throw new Error('exiting parenthesizes without having opened them; not supported'); // fail fast for unsupported case
      areInsideParensDepth -= 1; // mark that we've dropped off one parens now
    }

    // 2. cast the char, considering whether we're in parens or not
    const castedChar = areInsideParens() && char === ',' ? '|' : char; // cast "," if we are inside parens

    // 3. add casted char to string
    castedString += castedChar;
  });

  // return the casted string
  return castedString;
};
