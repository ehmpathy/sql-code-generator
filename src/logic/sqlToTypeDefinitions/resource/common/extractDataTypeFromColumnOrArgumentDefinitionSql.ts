import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://www.postgresql.org/docs/9.5/datatype-character.html
const postgresStringTypes = [
  'CHARACTER',
  'CHAR',
  'CHARACTER VARYING',
  'VARCHAR',
  'TEXT',
  'UUID', // this too: https://www.postgresql.org/docs/9.5/datatype-uuid.html
];

// combine the types so we can search over the union
const dbStringTypes = [...mysqlStringTypes, ...postgresStringTypes];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://www.postgresql.org/docs/9.5/datatype-numeric.html
const postgresNumberTypes = [
  'SMALLINT',
  'INT2',
  'INTEGER',
  'INT',
  'INT4',
  'BIGINT',
  'INT8',
  'DECIMAL',
  'NUMERIC',
  'REAL',
  'DOUBLE PRECISION',
  'FLOAT8',
  'SMALLSERIAL',
  'SERIAL2',
  'SERIAL',
  'SERIAL4',
  'BIGSERIAL',
  'SERIAL8',
];

// combine the types so we can search over the union
const dbNumberTypes = [...mysqlNumberTypes, ...postgresNumberTypes];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://www.postgresql.org/docs/9.5/datatype-datetime.html
const postgresDateTypes = [
  'TIMESTAMP',
  'TIMESTAMPTZ',
  'TIMESTAMP WITH TIMEZONE',
  'DATE',
  'TIME',
  'TIMETZ',
  'TIME WITH TIMEZONE',
];

// combine the types so we can search over the union
const dbDateTypes = [...mysqlDateTypes, ...postgresDateTypes];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

// https://www.postgresql.org/docs/9.5/datatype-binary.html
const postgresBinaryTypes = ['BYTEA'];

// combine the types so we can search over the union
const dbBinaryTypes = [...mysqlBinaryTypes, ...postgresBinaryTypes];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (dbStringTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.STRING;
  if (dbNumberTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.NUMBER;
  if (dbDateTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.DATE;
  if (dbBinaryTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
