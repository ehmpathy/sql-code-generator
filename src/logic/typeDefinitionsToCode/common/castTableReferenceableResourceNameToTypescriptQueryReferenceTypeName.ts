import { ResourceType } from '../../../model';
import { castResourceNameToTypescriptTypeName } from './castResourceNameToTypescriptTypeName';

export const castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  // 0. throw error if not view or table, only view and table are possible table references
  if (![ResourceType.TABLE, ResourceType.VIEW].includes(resourceType)) {
    throw new Error('not a table referenceable query');
  }

  // 1. get the regular typescript type name
  const typescriptTypeName = castResourceNameToTypescriptTypeName({ name, resourceType });

  // 2. return the referenceable name
  if (resourceType === ResourceType.TABLE) return typescriptTypeName.replace(/^SqlTable/, 'SqlTableOrView');
  if (resourceType === ResourceType.VIEW) return typescriptTypeName.replace(/^SqlView/, 'SqlTableOrView');
  throw new Error('unexpected'); // this should never occur, but fail fast
};
