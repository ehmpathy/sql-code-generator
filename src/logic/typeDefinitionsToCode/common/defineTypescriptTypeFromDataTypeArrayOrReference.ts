import {
  DataType,
  TypeDefinitionReference,
  TypeDefinition,
  TypeDefinitionOfQueryTableReference,
} from '../../../domain';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  plural,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  plural: boolean;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    const defined = defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
    if (plural) return `${defined}[]`;
    return defined;
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
