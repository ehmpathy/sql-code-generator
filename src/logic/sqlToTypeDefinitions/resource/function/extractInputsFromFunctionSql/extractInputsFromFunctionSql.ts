import strip from 'sql-strip-comments';

import { castCommasInParensToPipesForTokenSafety } from '../../common/castCommasInParensToPipesForTokenSafety';
import { extractTypeDefinitionFromFunctionInputSql } from './extractTypeDefinitionFromFunctionInputSql';

export const extractInputsFromFunctionSql = ({ sql }: { sql: string }) => {
  // 0. strip comments
  const strippedSql: string = strip(sql);

  // 1. grab the insides of the inputs definition
  const sqlBeforeReturns = strippedSql.split(/(?:returns|RETURNS)/)[0];
  const innerSqlAndAfter = sqlBeforeReturns
    .split('(')
    .slice(1)
    .join('('); // drop the part before the first '('
  const innerSql = innerSqlAndAfter
    .split(')')
    .slice(0, -1)
    .join(')'); // drop the part after the last ')'

  // 2. cast commas inside of parens into pipes, so that we treat them as unique tokens when splitting "property lines" by comma
  const parenTokenizedInnerSql = castCommasInParensToPipesForTokenSafety({ sql: innerSql }); // particularly useful for enum typedefs

  // 3. grab definition lines, by splitting out properties by commas
  const functionInputSqlSet = parenTokenizedInnerSql
    .split(',')
    .map((str) => str.trim()) // trim the line
    .filter((str) => !!str); // skip blank lines

  // 4. get column definition from each property
  return functionInputSqlSet.map((sql) => extractTypeDefinitionFromFunctionInputSql({ sql }));
};
