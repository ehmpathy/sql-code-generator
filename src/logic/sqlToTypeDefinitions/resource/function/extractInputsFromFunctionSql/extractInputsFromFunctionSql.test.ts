import { extractSqlFromFile } from '../../../../common/extractSqlFromFile';
import { extractInputsFromFunctionSql } from './extractInputsFromFunctionSql';

describe('extractInputsFromFunctionSql', () => {
  describe('mysql', () => {
    it('should extract the inputs accurately in this example', async () => {
      const defs = extractInputsFromFunctionSql({
        sql: await extractSqlFromFile({
          filePath: `${__dirname}/../../../../__test_assets__/functions/upsert_image.mysql.sql`,
        }),
      });
      expect(defs).toMatchSnapshot();
    });
    it('should extract the inputs accurately in this other example', async () => {
      const defs = extractInputsFromFunctionSql({
        sql: await extractSqlFromFile({
          filePath: `${__dirname}/../../../../__test_assets__/functions/upsert_suggestion.mysql.sql`,
        }),
      });
      expect(defs).toMatchSnapshot();
    });
    it('should extract the inputs accurately in this other example again', async () => {
      const defs = extractInputsFromFunctionSql({
        sql: await extractSqlFromFile({
          filePath: `${__dirname}/../../../../__test_assets__/functions/hash_string.mysql.sql`,
        }),
      });
      expect(defs).toMatchSnapshot();
    });
  });
  describe('postgres', () => {
    it('should extract the inputs accurately in this example', async () => {
      const defs = extractInputsFromFunctionSql({
        sql: await extractSqlFromFile({
          filePath: `${__dirname}/../../../../__test_assets__/functions/upsert_photo.postgres.sql`,
        }),
      });
      expect(defs).toMatchSnapshot();
    });
    it('should extract the inputs accurately when one of the inputs is an array type', async () => {
      const defs = extractInputsFromFunctionSql({
        sql: await extractSqlFromFile({
          filePath: `${__dirname}/../../../../__test_assets__/functions/upsert_job.postgres.sql`,
        }),
      });
      expect(defs).toMatchSnapshot();
    });
    it('should extract the inputs accurately there is no input args', async () => {
      const defs = extractInputsFromFunctionSql({
        sql: await extractSqlFromFile({
          filePath: `${__dirname}/../../../../__test_assets__/functions/get_answer_to_life.postgres.sql`,
        }),
      });
      expect(defs).toMatchSnapshot();
    });
  });
});
