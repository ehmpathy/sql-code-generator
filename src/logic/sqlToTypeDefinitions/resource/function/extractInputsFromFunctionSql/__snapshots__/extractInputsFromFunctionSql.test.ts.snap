// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractInputsFromFunctionSql mysql should extract the inputs accurately in this example 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_url",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_caption",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_credit",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_alt_text",
    "type": Array [
      "string",
      "null",
    ],
  },
]
`;

exports[`extractInputsFromFunctionSql mysql should extract the inputs accurately in this other example 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_suggestion_source",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_external_id",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_suggested_idea_id",
    "type": Array [
      "number",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_status",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_result",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_resultant_curated_idea_id",
    "type": Array [
      "number",
      "null",
    ],
  },
]
`;

exports[`extractInputsFromFunctionSql mysql should extract the inputs accurately in this other example again 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_array",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_delimiter",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_index",
    "type": Array [
      "number",
      "null",
    ],
  },
]
`;

exports[`extractInputsFromFunctionSql postgres should extract the inputs accurately in this example 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_url",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_description",
    "type": Array [
      "string",
      "null",
    ],
  },
]
`;

exports[`extractInputsFromFunctionSql postgres should extract the inputs accurately there is no input args 1`] = `Array []`;

exports[`extractInputsFromFunctionSql postgres should extract the inputs accurately when one of the inputs is an array type 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_uuid",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_title",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_status",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_description",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_location_id",
    "type": Array [
      "number",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_photo_ids",
    "type": Array [
      "number[]",
      "null",
    ],
  },
]
`;
